### PR TITLE
SDK feature parity

### PR DESCRIPTION
**Pull request by ftcvlad**
### Description of Changes

1) android: markAsRead and markAllInboxItemsAsRead. This adds features from [#55](https://github.com/Kumulos/KumulosSdkAndroid/pull/55)
2) android: Inbox enhancements from [#56](https://github.com/Kumulos/KumulosSdkAndroid/pull/56) . These include:
```
//add inboxUpdatedHandler
Kumulos.setOnInboxUpdatedHandler(() => {
      console.log('inbox updated');
});

//expose getInboxSummary method
Kumulos.getInboxSummary()
      .then((summary) => {
        console.log('total: '+summary.totalCount + ' unread: ' + summary.unreadCount);
      })
      .catch(() => {
        console.log('failed to get inbox summary!');
      });

//expose data and sentAt fields:
interface InAppInboxItem {
    .......
    isRead: boolean;
    sentAt?: string;
    data?: { [key: string]: any };
}
```
3) android: Tweaks from [#60](https://github.com/Kumulos/KumulosSdkAndroid/pull/60). imageUrl is added to InAppInboxItem and is always 300px wide, no base SDK image APIs exposed.

4) android: use latest SDK. This is a breaking change (release version 5.0.0) as need to add BoM dependency and switch to use androidx. This is done in the SDK. Users can override BoM version by adding build-extras.gradle to `/platforms/android/app` with following lines:
```
ext {
    KumulosFirebaseBoMVersion = "28.2.0"
}
```
Minimal version of Cordova becomes 9.0.0, minSdk 22.
5) android: remove jcenter
6) android,ios: dont block main thread, fix other compile and runtime warnings
7) ios: 1-3 changes
8) react: npm audit fix. Fix severe. To fix 4 moderate vulnerabilities update webpack, webpack-cli and typescript devDependencies to latest. Update webpack.config and fix other errors.
9) react: Use new channel subscription management endpoints using userIdentifier. No changes for users of the SDK
10) android: fix build.gradle file to properly find google-services.json and add ` com.google.gms.googleservices.GoogleServicesPlugin`. New cordova projects dont add `com.google.gms.google-services`, so, the error became apparent.

### Breaking Changes

- Use androidx instead of support library (if relevant)
- Minimal version of Cordova becomes 9.0.0, minSdk 22.

### Release Checklist

Prepare:

-   [x] `npm run dist` to produce minified artifacts
-   [x] Detail any breaking changes. Breaking changes require a new major version number
-   [x] Update type declarations in index.d.ts

Bump versions in:

-   [x] package.json
-   [x] plugin.xml
-   [x] src/ios/KumulosSDKPlugin.m
-   [x] src/android/.../KumulosInitProvider.java

Release:

-   [ ] Squash and merge to master
-   [ ] Delete branch once merged
-   [ ] Create tag from master matching chosen version
-   [ ] Fill out release notes
-   [ ] Run `npm publish`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,17 +28,16 @@
     ],
     "author": "Shaoyun Li",
     "license": "MIT",
-    "dependencies": {},
     "devDependencies": {
         "@types/cordova": "0.0.34",
         "babel-core": "^6.17.0",
         "babel-loader": "^7.0.0",
         "babel-plugin-syntax-dynamic-import": "^6.18.0",
         "babel-preset-env": "^1.7.0",
```
Comment (cgwyllie) [inline package.json:36]: Presume answer here is "yes", but are all plugins compatible with the webpack update? Suppose they're not webpack plugins.

File: src/android/KumulosSDKPlugin.java [hunk 2]
```diff
@@ -246,7 +260,7 @@ private void inAppUpdateConsent(JSONArray args, CallbackContext callbackContext)
     }
 
     private void inAppGetInboxItems(CallbackContext callbackContext) {
-        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US);
```
Comment (cgwyllie) [inline src/android/KumulosSDKPlugin.java:249]: Existing formatting bug?

Comment (cgwyllie) [reply src/android/KumulosSDKPlugin.java:249]: (any copy/paste from some existing SDK code in Android / other SDKs?)

File: src/android/KumulosSDKPlugin.java [hunk 3]
```diff
@@ -306,6 +326,7 @@ private void inAppPresentInboxMessage(JSONArray args, CallbackContext callbackCo
 
                 if (result == InboxMessagePresentationResult.PRESENTED) {
                     callbackContext.success();
+                    return;
```
Comment (cgwyllie) [inline src/android/KumulosSDKPlugin.java:329]: Oops ;)

File: src/android/kumulos.gradle [hunk 4]
```diff
@@ -39,7 +37,7 @@ buildscript {
 // to cdvPluginPostBuildExtras somewhere else, the plugin execution
 // will be skipped and project build will be successfull
 ext.postBuildExtras = {
-    if (file("./src/app/google-service.json").exists()) {
+    if (file("./google-services.json").exists()) {
```
Comment (cgwyllie) [inline src/android/kumulos.gradle:40]: Should we support both locations / any implications on integration guides?

IIRC google-services plugin looks (looked?) in both locations. Which is more "correct" / up to date? (thought it was under app/ but could be wrong)

File: src/core/client.ts [hunk 5]
```diff
@@ -27,6 +27,12 @@ export class Client {
         });
     }
 
+    public getUserIdentifier(): Promise<string> {
```
Comment (cgwyllie) [inline src/core/client.ts:30]: Naming consistency `getCurrentUserIden...` (what's most common)?

File: package.json (hunk 1 above)
Comment (ftcvlad) [reply package.json:36]: it managed to produce dist which is working when added to cordova. updated only the 3 that were required to update

File: src/core/index.ts [hunk 6]
```diff
@@ -386,7 +347,115 @@ const Kumulos = {
...
+                NativeModuleName,
+                'inAppDeleteMessageFromInbox',
+                [message.id]
+            );
+        });
+    },
+    /**
+     * Marks the given in-app inbox item as read
+     */
+    markAsRead: (message: InAppInboxItem): Promise<void> => {
```
Comment (cgwyllie) [inline src/core/index.ts:411]: Naming on these new methods, should we keep the consistent `inApp` prefix?

File: src/android/KumulosSDKPlugin.java (hunk 2 above)
Comment (ftcvlad) [reply src/android/KumulosSDKPlugin.java:249]: there was a warning in studio, base SDK has Z. Same in unity to come

File: src/android/KumulosSDKPlugin.java (hunk 3 above)
Comment (ftcvlad) [reply src/android/KumulosSDKPlugin.java:329]: yeah, found a lot of weird stuff in logs

**Comment by cgwyllie** [comment]
Regarding  the two breaking change items, can we explore how this affects things.

1) On BoM, is it still the case Cordova encourages you to treat the native project as a generated artefact? If so, manual modifications to the native project may not be desirable, suppose options are just bundling BoM version in our project.

2) androidx: what is the default with latest version of cordova?

3) androidx: is there not a specific way to "enable" the support in the cordova config.xml? (see release notes of the androidx support for cordova-android)

4) androidx: how might this affect docs/minimum supported cordova versions?

**Comment by cgwyllie** [comment]
Question on the change to the file name of the raven JS bundle, is it just changed due to webpack update and no further action necessary or is the loading affected and needing some more webpack config?

File: src/android/kumulos.gradle (hunk 4 above)
Comment (ftcvlad) [reply src/android/kumulos.gradle:40]: so, the story was: i created new cordova project, and it wouldnt register for push as this new cordova project doesnt add the plugin by default (but previous did for some reason).

so, here 
1) name was wrong, added 's' in the end (so, this was never found)
2) path file(.) _is_ platforms/android/app folder. src kind of shouldnt be there, as hint says `Resolves a file path relative to the project directory of this project`, but it also has app for some reason

Comment (cgwyllie) [reply src/android/kumulos.gradle:40]: o_o okay then.. Not sure how this was previously working, unless perhaps some other dep was applying / base project would have it somehow. Unless something else also changed.

File: src/core/client.ts (hunk 5 above)
Comment (ftcvlad) [reply src/core/client.ts]: yeah, it's current mb even everywhere. We dont have it in any docs as well. Updated

Comment (cgwyllie) [reply src/core/client.ts]: Cool. Shockingly consistent then!

Although boo on lack of docs. Will log a card to track.

File: src/core/index.ts (hunk 6 above)
Comment (ftcvlad) [reply src/core/index.ts]: yes, true. updated

File: src/android/kumulos.gradle (hunk 4 above)
Comment (ftcvlad) [reply src/android/kumulos.gradle:40]: new cordova project has 
`if (cdvHelpers.getConfigPreference('GradlePluginGoogleServicesEnabled', 'false').toBoolean()) {
    apply plugin: 'com.google.gms.google-services'
}`

previous one didnt have the conditional. dont know why it was adding, but looks it did

Comment (cgwyllie) [reply src/android/kumulos.gradle:40]: Ah ok, so really do we need this at all in our plugin code? I suppose it simplifies integration slightly, but perhaps we should just instead document enabling the setting in your config.xml if that is the official way now? (or can we declare it in our plugin.xml instead?)

Comment (ftcvlad) [reply src/android/kumulos.gradle:40]: could add to plugin.xml, but then it would be always added, even when no google-services.json present

**Pull request merged**